// @21st-sdk/node/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  AgentClient: () => AgentClient,
  AnClient: () => AnClient,
  VaultCoverageMissingError: () => VaultCoverageMissingError
});
module.exports = __toCommonJS(index_exports);

// src/client.ts
var DEFAULT_BASE_URL = "https://relay.an.dev";
var VaultCoverageMissingError = class extends Error {
  constructor(message, missing = []) {
    super(message);
    this.code = "vault_coverage_missing";
    this.name = "VaultCoverageMissingError";
    this.missing = missing;
  }
};
var AgentClient = class {
  constructor(config) {
    this.apiKey = config.apiKey;
    this.baseUrl = (config.baseUrl ?? DEFAULT_BASE_URL).replace(/\/$/, "");
    this.sandboxes = new SandboxesResource(this);
    this.threads = new ThreadsResource(this);
    this.tokens = new TokensResource(this);
  }
  /** @internal */
  async _request(path, init) {
    const url = `${this.baseUrl}${path}`;
    const res = await fetch(url, {
      ...init,
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.apiKey}`,
        ...init?.headers
      }
    });
    if (!res.ok) {
      const body = await res.json().catch(() => ({}));
      const msg = body.error?.message ?? `Request failed: ${res.status}`;
      if (body.error?.code === "vault_coverage_missing") {
        throw new VaultCoverageMissingError(msg, body.error.missing ?? []);
      }
      throw new Error(msg);
    }
    return res;
  }
  /** @internal */
  async _fetch(path, init) {
    const res = await this._request(path, init);
    if (res.status === 204) return void 0;
    return res.json();
  }
  /** @internal */
  _getBaseUrl() {
    return this.baseUrl;
  }
};
var FilesResource = class {
  constructor(client) {
    this.client = client;
  }
  async write(params) {
    return this.client._fetch(`/v1/sandboxes/${params.sandboxId}/files`, {
      method: "POST",
      body: JSON.stringify({ files: params.files })
    });
  }
  async read(params) {
    const encodedPath = encodeURIComponent(params.path);
    return this.client._fetch(
      `/v1/sandboxes/${params.sandboxId}/files?path=${encodedPath}`
    );
  }
  async list(params) {
    const query = new URLSearchParams({ path: params.path });
    if (params.depth !== void 0) {
      query.set("depth", String(params.depth));
    }
    return this.client._fetch(
      `/v1/sandboxes/${params.sandboxId}/files/list?${query.toString()}`
    );
  }
  async getInfo(params) {
    const encodedPath = encodeURIComponent(params.path);
    return this.client._fetch(
      `/v1/sandboxes/${params.sandboxId}/files/info?path=${encodedPath}`
    );
  }
  async exists(params) {
    const encodedPath = encodeURIComponent(params.path);
    return this.client._fetch(
      `/v1/sandboxes/${params.sandboxId}/files/exists?path=${encodedPath}`
    );
  }
  async makeDir(params) {
    return this.client._fetch(
      `/v1/sandboxes/${params.sandboxId}/files/mkdir`,
      {
        method: "POST",
        body: JSON.stringify({ path: params.path })
      }
    );
  }
  async rename(params) {
    return this.client._fetch(
      `/v1/sandboxes/${params.sandboxId}/files/rename`,
      {
        method: "POST",
        body: JSON.stringify({
          oldPath: params.oldPath,
          newPath: params.newPath
        })
      }
    );
  }
  async remove(params) {
    const encodedPath = encodeURIComponent(params.path);
    return this.client._fetch(
      `/v1/sandboxes/${params.sandboxId}/files?path=${encodedPath}`,
      { method: "DELETE" }
    );
  }
};
var GitResource = class {
  constructor(client) {
    this.client = client;
  }
  /** @deprecated Use `sandboxes.exec({ sandboxId, command: 'git clone ...' })` instead. */
  async clone(params) {
    console.warn(
      "[an-sdk] sandboxes.git.clone() is deprecated and will be removed in a future release. Use sandboxes.exec() instead."
    );
    return this.client._fetch(
      `/v1/sandboxes/${params.sandboxId}/git/clone`,
      {
        method: "POST",
        body: JSON.stringify({
          url: params.url,
          ...params.path && { path: params.path },
          ...params.token && { token: params.token },
          ...params.depth && { depth: params.depth }
        })
      }
    );
  }
};
var SandboxesResource = class {
  constructor(client) {
    this.client = client;
    this.files = new FilesResource(client);
    this.git = new GitResource(client);
  }
  async create(params) {
    return this.client._fetch("/v1/sandboxes", {
      method: "POST",
      body: JSON.stringify({
        agent: params.agent,
        ...params.files && { files: params.files },
        ...params.envs && { envs: params.envs },
        ...params.setup && { setup: params.setup },
        ...params.timeoutMs !== void 0 && { timeoutMs: params.timeoutMs },
        ...params.networkAllowOut && {
          networkAllowOut: params.networkAllowOut
        },
        ...params.networkDenyOut && { networkDenyOut: params.networkDenyOut }
      })
    });
  }
  async get(sandboxId) {
    return this.client._fetch(`/v1/sandboxes/${sandboxId}`);
  }
  async delete(sandboxId) {
    return this.client._fetch(`/v1/sandboxes/${sandboxId}`, {
      method: "DELETE"
    });
  }
  async exec(params) {
    return this.client._fetch(
      `/v1/sandboxes/${params.sandboxId}/exec`,
      {
        method: "POST",
        body: JSON.stringify({
          command: params.command,
          ...params.cwd && { cwd: params.cwd },
          ...params.envs && { envs: params.envs },
          ...params.timeoutMs && { timeoutMs: params.timeoutMs }
        })
      }
    );
  }
};
var ThreadsResource = class {
  constructor(client) {
    this.client = client;
  }
  async list(params) {
    return this.client._fetch(
      `/v1/sandboxes/${params.sandboxId}/threads`
    );
  }
  async create(params) {
    return this.client._fetch(
      `/v1/sandboxes/${params.sandboxId}/threads`,
      {
        method: "POST",
        body: JSON.stringify({ name: params.name })
      }
    );
  }
  async get(params) {
    return this.client._fetch(
      `/v1/sandboxes/${params.sandboxId}/threads/${params.threadId}`
    );
  }
  async delete(params) {
    return this.client._fetch(
      `/v1/sandboxes/${params.sandboxId}/threads/${params.threadId}`,
      { method: "DELETE" }
    );
  }
  async run(params) {
    if (params.threadId && !params.sandboxId) {
      throw new Error("threadId requires sandboxId");
    }
    const encodedAgent = encodeURIComponent(params.agent);
    const sandboxId = params.sandboxId ?? (await this.client.sandboxes.create({ agent: params.agent })).id;
    const threadId = params.threadId ?? (await this.create({ sandboxId, name: params.name })).id;
    const vaultIds = normalizeVaultIds(params);
    if (params.mode === "background") {
      const response2 = await this.client._request(`/v1/chat/${encodedAgent}`, {
        method: "POST",
        body: JSON.stringify({
          messages: params.messages,
          sandboxId,
          threadId,
          mode: "background",
          ...vaultIds && { vaultIds },
          ...params.externalUserId && { externalUserId: params.externalUserId },
          ...params.options && { options: params.options }
        })
      });
      if (response2.status !== 202) {
        void response2.text().catch(() => {
        });
        return { sandboxId, threadId, status: "running" };
      }
      const result = await response2.json().catch(() => ({ status: "running" }));
      return { sandboxId, threadId, status: result.status ?? "running" };
    }
    const response = await this.client._request(`/v1/chat/${encodedAgent}`, {
      method: "POST",
      body: JSON.stringify({
        messages: params.messages,
        sandboxId,
        threadId,
        ...vaultIds && { vaultIds },
        ...params.externalUserId && { externalUserId: params.externalUserId },
        ...params.options && { options: params.options }
      })
    });
    return {
      sandboxId,
      threadId,
      response,
      resumeUrl: `${this.client._getBaseUrl()}/v1/chat/${encodedAgent}/${sandboxId}/stream`
    };
  }
};
var didWarnVaultIdDeprecated = false;
var didWarnVaultIdIgnored = false;
function warnOnce(kind, message) {
  if (kind === "deprecated") {
    if (didWarnVaultIdDeprecated) return;
    didWarnVaultIdDeprecated = true;
  } else {
    if (didWarnVaultIdIgnored) return;
    didWarnVaultIdIgnored = true;
  }
  console.warn(message);
}
function normalizeVaultIds(params) {
  if (params.vaultIds) {
    if (params.vaultId) {
      warnOnce("ignored", "vaultId was ignored because vaultIds was provided");
    }
    return [...new Set(params.vaultIds)];
  }
  if (params.vaultId) {
    warnOnce("deprecated", "vaultId is deprecated, use vaultIds: string[]");
    return [params.vaultId];
  }
  return void 0;
}
var TokensResource = class {
  constructor(client) {
    this.client = client;
  }
  async create(params = {}) {
    return this.client._fetch("/v1/tokens", {
      method: "POST",
      body: JSON.stringify({
        agents: params.agent ? [params.agent] : void 0,
        userId: params.userId,
        expiresIn: params.expiresIn ?? "1h"
      })
    });
  }
};
var AnClient = AgentClient;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AgentClient,
  AnClient,
  VaultCoverageMissingError
});
